### PR TITLE
Add rule discovery tunings

### PR DESCRIPTION
**Pull request by RomanSraj**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -19,4 +19,5 @@ joblib~=1.1.0
 click~=8.0.3
 setuptools~=57.0.0
 baycomp~=1.0.2
-tabulate~=0.8.9
\ No newline at end of file
+tabulate~=0.8.9
+xcsf~=1.2.5
```
Comment (heidmic) [inline requirements.txt:23]: Add a newline after this

File: runs/comparisons/decision_tree.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        # params.min_samples_leaf = trial.suggest_int('min_samples_leaf', 1, 4)
+        # params.min_weight_fraction_leaf = trial.suggest_float('min_weight_fraction_leaf', 0.0, 0.5)
+        params.max_features = trial.suggest_categorical('max_features', ["auto", "sqrt", "log2"])
+        # params.min_impurity_decrease = trial.suggest_float('min_impurity_decrease', 0.0, 0.5)
+        # params.ccp_alpha = trial.suggest_float('ccp_alpha', 0.0, 0.02)
+
+    experiment.with_tuning(optuna_objective, tuner=optuna_tuner)
+
+    # Create a nested experiment using the MAE.
+    mae_experiment = experiment.with_params({'criterion': 'absolute_error'})
```
Comment (heidmic) [inline runs/comparisons/decision_tree.py:79]: Why do we want to so this?

File: runs/comparisons/rf.py [hunk 3]
```diff
@@ -29,7 +34,7 @@
         X_train=X_train,
         y_train=y_train,
         random_state=random_state,
-        n_calls=10,
+        n_calls=1000,
```
Comment (heidmic) [inline runs/comparisons/rf.py:37]: Might be the phone but why is there an indent here?

File: runs/comparisons/rf.py [hunk 4]
```diff
@@ -39,13 +44,14 @@
         X_train=X_train,
         y_train=y_train,
         random_state=random_state,
-        n_calls=20,
+        n_calls=1000,
```
Comment (heidmic) [inline runs/comparisons/rf.py:47]: Why are we using skopt and optuna based tuners?

File: runs/comparisons/rf.py [hunk 5]
```diff
@@ -39,13 +44,14 @@
...
         verbose=10,
         n_jobs_cv=4,
     )
 
     # Create the base experiment, using some default tuner
-    experiment = Experiment(name='Random Forest', tuner=default_tuner, verbose=10)
+    experiment_name = f'Random Forest {job_id} {problem}'
+    experiment = Experiment(name=experiment_name, tuner=default_tuner, verbose=10)
```
Comment (heidmic) [inline runs/comparisons/rf.py:54]: What is the default_tuner? Shouldn't this be one of the ones we defined above?

File: runs/comparisons/xcsf_final.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
+import mlflow
```
Comment (heidmic) [inline runs/comparisons/xcsf_final.py:1]: This file contains a lot of code duplication

File: runs/rule_discovery/evolution_strategy.py [hunk 7]
```diff
@@ -66,7 +66,7 @@ def run(problem: str, job_id: str):
         cv=4,
         n_jobs_cv=4,
         n_jobs=4,
-        n_calls=10_000,
+        n_calls=1000,
         timeout=24 * 60 * 60,  # 24 hours
```
Comment (heidmic) [inline runs/rule_discovery/evolution_strategy.py:70]: Budget should be 72?

**Review by heidmic — CHANGES_REQUESTED** [review]
Did not manage to review XCSF in detail yet

File: runs/comparisons/decision_tree.py (hunk 2 above)
Comment (RomanSraj) [reply runs/comparisons/decision_tree.py:79]: I don't know the reason behind this. I just moved this file and didn't change anything there yet. I can investigate this in a different PR

File: runs/comparisons/rf.py (hunk 3 above)
Comment (RomanSraj) [reply runs/comparisons/rf.py:37]: I can't see an indent, so it's probably the phone

File: runs/comparisons/rf.py (hunk 4 above)
Comment (RomanSraj) [reply runs/comparisons/rf.py:47]: I don't know what the reason behind using both tuners was. If we only need one, I can remove the other (I'm guessing optuna is the one we prefer)

File: runs/comparisons/rf.py (hunk 5 above)
Comment (RomanSraj) [reply runs/comparisons/rf.py:54]: default_tuner is SkoptTuner. If we don't need that one I can remove it

File: runs/comparisons/decision_tree.py (hunk 2 above)
Comment (heidmic) [reply runs/comparisons/decision_tree.py]: We should at least find out what it does

File: runs/comparisons/rf.py (hunk 4 above)
Comment (heidmic) [reply runs/comparisons/rf.py:40]: as far as I see below we only call optuna, so we should definitely remove the other

File: runs/comparisons/rf.py (hunk 5 above)
Comment (heidmic) [reply runs/comparisons/rf.py]: We should use optuna here

File: runs/comparisons/xcsf_tuning.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+        xcs.action("integer")  # (dummy) integer actions
+
+        args = {
+            "min": -1,  # minimum value of a lower bound
+            "max": 1,  # maximum value of an upper bound
+            "spread_min": 0.1,  # minimum initial spread
+            "eta":
+                0,  # disable gradient descent of centers towards matched input mean
+        }
+        xcs.condition("hyperrectangle_csr", args)
```
Comment (heidmic) [inline runs/comparisons/xcsf_tuning.py]: change

File: runs/comparisons/xcsf_final.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+        xcs.action("integer")  # (dummy) integer actions
+
+        args = {
+            "min": -1,  # minimum value of a lower bound
+            "max": 1,  # maximum value of an upper bound
+            "spread_min": 0.1,  # minimum initial spread
+            "eta":
+                0,  # disable gradient descent of centers towards matched input mean
+        }
+        xcs.condition("hyperrectangle_csr", args)
```
Comment (heidmic) [inline runs/comparisons/xcsf_final.py]: change

**Review by heidmic — CHANGES_REQUESTED** [review]

File: runs/comparisons/rf.py (hunk 4 above)
Comment (RomanSraj) [reply runs/comparisons/rf.py:40]: Removed

File: runs/comparisons/rf.py (hunk 5 above)
Comment (RomanSraj) [reply runs/comparisons/rf.py]: SkoptTuner removed

File: runs/comparisons/xcsf_final.py (hunk 6 above)
Comment (RomanSraj) [reply runs/comparisons/xcsf_final.py]: xcsf moved to its own branch (with the rebased master https://github.com/heidmic/suprb-experimentation/tree/xcsf_experiment)

File: runs/comparisons/xcsf_tuning.py (hunk 8 above)
Comment (RomanSraj) [reply runs/comparisons/xcsf_tuning.py]: Moved to its own branch https://github.com/heidmic/suprb-experimentation/tree/xcsf_experiment

File: runs/comparisons/xcsf_final.py (hunk 9 above)
Comment (RomanSraj) [reply runs/comparisons/xcsf_final.py]: Moved to its own branch https://github.com/heidmic/suprb-experimentation/tree/xcsf_experiment

File: runs/rule_discovery/evolution_strategy.py [hunk 10]
```diff
@@ -34,9 +34,8 @@ def load_dataset(name: str, **kwargs) -> tuple[np.ndarray, np.ndarray]:
 
 @click.command()
 @click.option('-p', '--problem', type=click.STRING, default='airfoil_self_noise')
-@click.option('-j', '--job_id', type=click.STRING, default='NA')
-def run(problem: str, job_id: str):
-    print(f"Problem is {problem}, with job id {job_id}")
+def run(problem: str):
+    print(f"Problem is {problem}")
```
Comment (heidmic) [inline runs/rule_discovery/evolution_strategy.py:38]: Why did you remove the jobid again?

File: slurm/default.sbatch [hunk 11]
```diff
@@ -1,10 +1,9 @@
 #!/usr/bin/env bash
-#SBATCH --time=35:00:00
-#SBATCH --partition=cpu-prio
+#SBATCH --time=96:00:00
+#SBATCH --partition=gpu
 #SBATCH --output=output/output-%j.txt
-#SBATCH --cpus-per-task=15
-#SBATCH --mem-per-cpu=1G
+#SBATCH --cpus-per-task=2
+#SBATCH --mem-per-cpu=4G
```
Comment (heidmic) [inline slurm/default.sbatch:6]: partition should default to cpu with 15 cores and 30 GB of memory (total). max time is okay for a default, imho.

**Review by heidmic — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 12]
```diff
@@ -19,4 +19,5 @@ joblib~=1.1.0
 click~=8.0.3
 setuptools~=57.0.0
 baycomp~=1.0.2
-tabulate~=0.8.9
\ No newline at end of file
+tabulate~=0.8.9
+xcsf~=1.1.6
```
Comment (heidmic) [inline requirements.txt:23]: remove from this branch

File: runs/comparisons/rf.py [hunk 13]
```diff
@@ -3,50 +3,45 @@
...
 from sklearn.model_selection import train_test_split
-from sklearn.utils import Bunch
+from sklearn.utils import Bunch, shuffle
 from skopt.space import Integer, Real
+from sklearn.model_selection import ShuffleSplit
+
 
 from experiments import Experiment
-from experiments.evaluation import CrossValidateTest
+from experiments.evaluation import CrossValidate
```
Comment (heidmic) [inline runs/comparisons/rf.py:12]: what is the difference between the two?

File: runs/comparisons/rf.py [hunk 14]
```diff
@@ -3,50 +3,45 @@
...
+@click.command()
+@click.option('-p', '--problem', type=click.STRING, default='airfoil_self_noise')
+@click.option('-j', '--job_id', type=click.STRING, default='NA')
+def run(problem: str, job_id: str):
     random_state = 42
 
     X, y = load_airfoil_self_noise()
     X, y = scale_X_y(X, y)
-    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=random_state)
+    X, y = shuffle(X, y, random_state=random_state)
```
Comment (heidmic) [inline runs/comparisons/rf.py:30]: what does shuffle do?

File: runs/comparisons/rf.py [hunk 15]
```diff
@@ -3,50 +3,45 @@
...
-        X_train=X_train,
-        y_train=y_train,
         random_state=random_state,
-        n_calls=20,
-        verbose=10,
+        cv=4,
         n_jobs_cv=4,
+        n_jobs=4,
+        n_calls=1000,
+        timeout=72 * 60 * 60,  # 24 hours
```
Comment (heidmic) [inline runs/comparisons/rf.py:41]: update comment

File: runs/comparisons/suprb_tuning.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    )
+
+    @param_space()
+    def suprb_ES_GA_space(trial: Trial, params: Bunch):
+        # ES
+        sigma_space = [0, np.sqrt(X.shape[1])]
+
+        params.rule_generation__mutation__sigma = trial.suggest_float(
+            'rule_generation__mutation__sigma', *sigma_space)
+        params.rule_generation__delay = trial.suggest_int('rule_generation__delay', 10, 100)
```
Comment (heidmic) [inline runs/comparisons/suprb_tuning.py:82]: should be conditional on `operator` being '&'. Otherwise, this has no effect and confuses the tuner

File: runs/rule_discovery/evolution_strategy.py (hunk 10 above)
Comment (RomanSraj) [reply runs/rule_discovery/evolution_strategy.py:38]: reset to original state

**Review by heidmic — APPROVED** [review]
